### ISSUE TITLE
utils.Sequence iterates forever

### ISSUE THREAD

**Issue by MrSnappingTurtle** [keras-team/keras#11438]
As per [Goodword's question](https://stackoverflow.com/questions/52267297/why-does-an-instance-of-the-keras-sequence-iterate-forever), the `keras.utils.Sequence` class seems to iterate forever instead of stopping after a full epoch. To duplicate, run the following snippet. 

```
class InfiniteGenerator(object):
    def __init__(self, alist):
        self.alist = alist

    def __getitem__(self, idx):
        return self.alist[idx]

    def __len__(self):
        return len(self.alist)

    def __iter__(self):
        for item in (self[i] for i in range(len(self))):
            yield item

from keras.utils import Sequence

class KGen(Sequence):
    def __init__(self, alist):
        self.alist = alist

    def __getitem__(self, idx):
        return self.alist[idx]

    def __len__(self):
        return len(self.alist)


if __name__ ==  '__main__':
    ig = InfiniteGenerator(list(range(4)))
    for item in ig:
        print(item)


    print('now trying second iterator')

    import time
    time.sleep(1)

    kg = KGen(list(range(4)))
    for item in kg:
        print(item)
```
The output will be as follows:

```
0
1
2
3
now trying second iterator
0
1
2
3
0
1
2
3
0
1
2
3
and so on ...
```

Tested on

* Ubuntu 18.04
* Keras 2.2.4
* Tensorflow 1.11.0

**Comment by jacquerie** [comment]
FWIW, I tried this code on Python 2.7.15rc1, Python 3.6.6, and Conda Python 3.7 and in all cases I see it terminating...

**Comment by MrSnappingTurtle** [comment]
I'll try matching my Python environment to yours and if I can still replicate.

**Comment by MFreidank** [comment]
I can reproduce your behavior, but I am not sure this is actually an issue. What indicates to you that it should actually terminate after an epoch?

**Comment by MrSnappingTurtle** [comment]
@MFreidank From my understanding (and apologies in advance if I got something wrong, I'm still new to Python), [since 2.2.3, the generator is no longer infinite](https://github.com/keras-team/keras/commit/5027630fa41f499a9226a8f9d952ceabf2c247aa#diff-ba9d38600a2df565e5ae8757eb2b1b35), so I believe it should only loop once.

@jacquerie So I tried using non-conda Python 3.6.4 (installed through pyenv) and it works fine now. Going to try and use conda again to see if the issue returns... @MFreidank, may I ask what Python environment you're using?

As a side note, if this is in fact a Python issue and not a Keras one, would the right thing for me to do be to close this issue?
